DESIGN.md

```diff
@@ -0,0 +1,211 @@
+    - *worker*: The actual Linux service to manage jobs. Using a service is convenient because it provides 
+    automatic restarts if the service crashes due to somebody running something that consumes enormous RAM or 
+    something.
+        - Exposes the methods for Start, Stop, Status, Output, List via RPC
+- *Makefile*: will build one or all of the 3 executables
+
+
+### Worker
+The job worker will make use of `net/rpc` for communicating with the HTTP server. The package is in feature lock, but 
+provides an easy interface for IPC and development is still current. There have been new commits to the package that
+will be included in the Go 1.17 release. Using Unix domain sockets will allow pretty low latency and is supported by 
+net/rpc. Better performance could be possible, but won't be necessary for this example. 
+
+The `os/exec` package allows capturing stdout/stderr directly and store it in the job. Not memory efficient to keep it 
+all in RAM, but quick to implement. Storing stdout and stderr will be done separately to allow the user to get get 
+them as two separate fields when retreiving output.
+
+Worker will initialize logging, initialize worker hub, and intialize an RPC listener. Worker hub will contain a list
+of all workers.
+
+Job type will include the job owner id, the command that started it, times for different events such as start 
+and finish, byte slices of stderr and stdout and other relevant information required by the API and CLI. Listener
+will make a few methods available via RPC for use by the API and CLI clients. The List method will only list jobs
+owned by the users passed in as an argument, or all if there are no users passed in.
+
+Each job will be protected by a mutex to prevent conflicting commands.
+
+The hub will have a map of jobID to jobs. The length of this map will be used to generate the next jobID. The map will 
+also be protected by a mutex to prevent simultaneous access.
+
+### API
+Will use the `net/http` package for ease of use and fast development. Server will be configured for mTLS with TLS
+version 1.3 only. The HTTP server will have a locally generated CA cert to validate the client TLS certificates, and 
+middleware will perform the authorization based on the certificate's Subject field before passing requests off to the 
+handlers. As there are only 5 endpoints, no third-party mux will be used and path variables will be parsed manually.
+
+### CLI
+The CLI will be built with the `github.com/spf13/cobra` library for fast development. Server address and credentials 
+will be hard-coded for simplicity. The CLI will communicate with the HTTP server configured for TLS. THe client will
+have the same CA cert in its cert pool for validating server certificates.
+
+#### Examples
+```
+> alps start `cd /var/log && ls -lh`
+Started Job 123 for user@example.com
+> alps stop 123
+Stopped Job 123. Output will be available for 1 hour.
```

Comment:
Why just 1 hour?

Reply:
That was arbitrarily chosen when I was planning on cleaning up finished jobs to save memory. @smallinsky recommended removing that from scope, so job output will be available indefinitely.